### ISSUE TITLE
Parse `expand` of dim=1 to be a broadcast

### ISSUE THREAD

**Issue by kevinstephano** [csarofeen/pytorch#1507]
### 🚀 The feature, motivation and pitch

In the backward pass of python written operations like Layer Norm, we see that the multiplication of reduction output in the forward pass results in reduction followed by an expand instead of broadcast in the backward pass.  Therefore, it would be nice to treat expands of dimensions of size 1 as broadcasts.

Here is a simple example:

```
import torch

class Fusion(torch.nn.Module) :
    def __init__(self) :
        super(Fusion, self).__init__()

    def forward(self, inputs) :
        out1 = inputs * 2.                                                                                                                                                                                                                           
        out2 = out1.sum(dim=-1, keepdim=True)                                                                                                                                                                                                        
        out3 = out2.expand((2, 4, 6))                                                                                                                                                                                                                                                                                                                                                                                                                                             
        out4 = out3 * 3.                                                                                                                                                                                                                             
        out5 = out4 * 5.                                                                                                                                                                                                                             
        return out5                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                     
 
if __name__ == "__main__" :
    inputs = torch.randn(2, 4, 6, device='cuda')                                                                                                                                                                                                     
                                                                                                                                                                                                                                                     
    model = torch.jit.script(Fusion())
                                                                                                                                                                                                                                                     
    with torch.jit.fuser('fuser2') :
        for _ in range(5) :                                                                                                                                                                                                                          
            out = model(inputs) 
```

This results in two fusion groups whereas removing the expand and leaving the broadcast results in one.  I am not sure if this is also something Autograd could avoid.

```
$ PYTORCH_JIT_LOG_LEVEL=graph_fuser python test.py                                                                        
[DUMP graph_fuser.cpp:2307] Before Fusion:                                                                                                                                                                                                           
[DUMP graph_fuser.cpp:2307] graph(%self : __torch__.Fusion,                                                               
[DUMP graph_fuser.cpp:2307]       %inputs.1 : Tensor):                                                                                                                                                                                               
[DUMP graph_fuser.cpp:2307]   %9 : float = prim::Constant[value=5.]() # test.py:12:22                                     
[DUMP graph_fuser.cpp:2307]   %8 : float = prim::Constant[value=3.]() # test.py:11:22                                     
[DUMP graph_fuser.cpp:2307]   %7 : float = prim::Constant[value=2.]() # test.py:8:24                                      
[DUMP graph_fuser.cpp:2307]   %6 : bool = prim::Constant[value=1]() # test.py:9:40                                                                                                                                                                   
[DUMP graph_fuser.cpp:2307]   %5 : NoneType = prim::Constant()                                                                                                                                                                                       
[DUMP graph_fuser.cpp:2307]   %4 : bool = prim::Constant[value=0]()                                                       
[DUMP graph_fuser.cpp:2307]   %3 : int[] = prim::Constant[value=[-1]]()                                                   
[DUMP graph_fuser.cpp:2307]   %2 : int[] = prim::Constant[value=[2, 4, 6]]()                                              
[DUMP graph_fuser.cpp:2307]   %10 : Tensor = prim::profile[profiled_type=Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0), seen_none=0](%inputs.1)                                                                                 
[DUMP graph_fuser.cpp:2307]   %out1.1 : Tensor = aten::mul(%10, %7) # test.py:8:15                                        
[DUMP graph_fuser.cpp:2307]   %12 : Tensor = prim::profile[profiled_type=Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0), seen_none=0](%out1.1)                                                                                   
[DUMP graph_fuser.cpp:2307]   %out2.1 : Tensor = aten::sum(%12, %3, %6, %5) # test.py:9:15                                
[DUMP graph_fuser.cpp:2307]   %14 : Tensor = prim::profile[profiled_type=Float(2, 4, 1, strides=[4, 1, 1], requires_grad=0, device=cuda:0), seen_none=0](%out2.1)                                                                                    
[DUMP graph_fuser.cpp:2307]   %out3.1 : Tensor = aten::expand(%14, %2, %4) # test.py:10:15                                                                                                                                                           
[DUMP graph_fuser.cpp:2307]   %16 : Tensor = prim::profile[profiled_type=Float(2, 4, 6, strides=[4, 1, 0], requires_grad=0, device=cuda:0), seen_none=0](%out3.1)                                                                                    
[DUMP graph_fuser.cpp:2307]   %out4.1 : Tensor = aten::mul(%16, %8) # test.py:11:15  
[DUMP graph_fuser.cpp:2307]   %18 : Tensor = prim::profile[profiled_type=Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0), seen_none=0](%out4.1)                                                                                   
[DUMP graph_fuser.cpp:2307]   %out5.1 : Tensor = aten::mul(%18, %9) # test.py:12:15
[DUMP graph_fuser.cpp:2307]   %20 : Tensor = prim::profile[profiled_type=Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0), seen_none=0](%out5.1)                                                                                   
[DUMP graph_fuser.cpp:2307]   return (%20)                                                                                                                                                                                                           
[DUMP graph_fuser.cpp:2385] Before Compilation:                                                                                                                                                                                                      
[DUMP graph_fuser.cpp:2385] graph(%self : __torch__.Fusion,                                                                                                                                                                                          
[DUMP graph_fuser.cpp:2385]       %inputs.1 : Tensor):                                                                                                                                                                                               
[DUMP graph_fuser.cpp:2385]   %4 : bool = prim::Constant[value=0]()                                                                                                                                                                                  
[DUMP graph_fuser.cpp:2385]   %2 : int[] = prim::Constant[value=[2, 4, 6]]()                                              
[DUMP graph_fuser.cpp:2385]   %31 : bool = prim::CudaFusionGuard[types=[Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0)]](%inputs.1)                                                                                              
[DUMP graph_fuser.cpp:2385]   %30 : Tensor = prim::If(%31)                                                                
[DUMP graph_fuser.cpp:2385]     block0():                                                                                                                                                                                                            
[DUMP graph_fuser.cpp:2385]       %out2.4 : Tensor = prim::CudaFusionGroup_0(%inputs.1)                                                                                                                                                              
[DUMP graph_fuser.cpp:2385]       -> (%out2.4)                                                                            
[DUMP graph_fuser.cpp:2385]     block1():                                                                                 
[DUMP graph_fuser.cpp:2385]       %out2.1 : Tensor = prim::FallbackGraph_1(%inputs.1)                                     
[DUMP graph_fuser.cpp:2385]       -> (%out2.1)                                                                                                                                                                                                       
[DUMP graph_fuser.cpp:2385]   %out3.1 : Tensor = aten::expand(%30, %2, %4) # test.py:10:15                                                                                                                                                           
[DUMP graph_fuser.cpp:2385]   %40 : bool = prim::CudaFusionGuard[types=[Float(2, 4, 6, strides=[4, 1, 0], requires_grad=0, device=cuda:0)]](%out3.1)                                                                                                 
[DUMP graph_fuser.cpp:2385]   %39 : Tensor = prim::If(%40)                                                                                                                                                                                           
[DUMP graph_fuser.cpp:2385]     block0():                                                                                 
[DUMP graph_fuser.cpp:2385]       %out5.4 : Tensor = prim::CudaFusionGroup_2(%out3.1)                                                                                                                                                                
[DUMP graph_fuser.cpp:2385]       -> (%out5.4)                                                                            
[DUMP graph_fuser.cpp:2385]     block1():                                                                                 
[DUMP graph_fuser.cpp:2385]       %out5.1 : Tensor = prim::FallbackGraph_3(%out3.1)                                       
[DUMP graph_fuser.cpp:2385]       -> (%out5.1)                                                                                                                                                                                                       
[DUMP graph_fuser.cpp:2385]   return (%39)                                                                                
[DUMP graph_fuser.cpp:2385] with prim::CudaFusionGroup_0 = graph(%5 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0)):                                                                                                           
[DUMP graph_fuser.cpp:2385]   %3 : NoneType = prim::Constant()                                                            
[DUMP graph_fuser.cpp:2385]   %2 : bool = prim::Constant[value=1]() # test.py:9:40                                                                                                                                                                   
[DUMP graph_fuser.cpp:2385]   %1 : int[] = prim::Constant[value=[-1]]()                                                   
[DUMP graph_fuser.cpp:2385]   %6 : float = prim::Constant[value=2.]() # test.py:8:24                                                                                                                                                                 
[DUMP graph_fuser.cpp:2385]   %out1.1 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0) = aten::mul(%5, %6) # test.py:8:15                                                                                                        
[DUMP graph_fuser.cpp:2385]   %out2.1 : Float(2, 4, 1, strides=[4, 1, 1], requires_grad=0, device=cuda:0) = aten::sum(%out1.1, %1, %2, %3) # test.py:9:15                                                                                            
[DUMP graph_fuser.cpp:2385]   return (%out2.1)                                                                            
[DUMP graph_fuser.cpp:2385] with prim::FallbackGraph_1 = graph(%inputs.1 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0)):                                                                                                      
[DUMP graph_fuser.cpp:2385]   %1 : NoneType = prim::Constant()                                                            
[DUMP graph_fuser.cpp:2385]   %2 : bool = prim::Constant[value=1]() # test.py:9:40                                        
[DUMP graph_fuser.cpp:2385]   %3 : int[] = prim::Constant[value=[-1]]()                                                   
[DUMP graph_fuser.cpp:2385]   %4 : float = prim::Constant[value=2.]() # test.py:8:24                                      
[DUMP graph_fuser.cpp:2385]   %out1.1 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0) = aten::mul(%inputs.1, %4) # test.py:8:15                                                                                                 
[DUMP graph_fuser.cpp:2385]   %out2.1 : Float(2, 4, 1, strides=[4, 1, 1], requires_grad=0, device=cuda:0) = aten::sum(%out1.1, %3, %2, %1) # test.py:9:15                                                                                            
[DUMP graph_fuser.cpp:2385]   return (%out2.1)                                                                            
[DUMP graph_fuser.cpp:2385] with prim::CudaFusionGroup_2 = graph(%3 : Float(2, 4, 6, strides=[4, 1, 0], requires_grad=0, device=cuda:0)):                                                                                                            
[DUMP graph_fuser.cpp:2385]   %1 : float = prim::Constant[value=5.]() # test.py:12:22                                     
[DUMP graph_fuser.cpp:2385]   %4 : float = prim::Constant[value=3.]() # test.py:11:22                                     
[DUMP graph_fuser.cpp:2385]   %out4.1 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0) = aten::mul(%3, %4) # test.py:11:15                                                                                                       
[DUMP graph_fuser.cpp:2385]   %out5.1 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0) = aten::mul(%out4.1, %1) # test.py:12:15                                                                                                  
[DUMP graph_fuser.cpp:2385]   return (%out5.1)                                                                            
[DUMP graph_fuser.cpp:2385] with prim::FallbackGraph_3 = graph(%out3.1 : Float(2, 4, 6, strides=[4, 1, 0], requires_grad=0, device=cuda:0)):
[DUMP graph_fuser.cpp:2385]   %1 : float = prim::Constant[value=5.]() # test.py:12:22
[DUMP graph_fuser.cpp:2385]   %2 : float = prim::Constant[value=3.]() # test.py:11:22
[DUMP graph_fuser.cpp:2385]   %out4.1 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0) = aten::mul(%out3.1, %2) # test.py:11:15
[DUMP graph_fuser.cpp:2385]   %out5.1 : Float(2, 4, 6, strides=[24, 6, 1], requires_grad=0, device=cuda:0) = aten::mul(%out4.1, %1) # test.py:12:15
[DUMP graph_fuser.cpp:2385]   return (%out5.1)            
```

### Alternatives

_No response_

### Additional context

_No response_

**Comment by jjsjann123** [comment]
Quick summary on the request:

1. We would only support size-1 expand can won't allow expanding a dimension to a real size. Not sure how nvfuser would support that, as we can't specify the broadcasted dimension to a concrete size naturally. We can probably fake it by some binary ops with empty tensors, but I think that's too far.
2. We do want to support dynamic sizes as well, so we'll need some guard logic similar to `view`
3. Another thing that we need to be careful with are aliases. Expand in eager does create an alias, so we'll be careful to NOT to output both tensors to avoid diverging on eager aliasing behavior.

**Comment by rdspring1** [comment]
1. Could we avoid fusing `expand` if it is a fusion output, since we will need to return an aliased tensor with stride 0?
2. The `CudaFusionGuard` should check that the expanded dimension size matches the resolved broadcast dimension size.
3. Treat `expand` as a no-op in the fusion because we're using expand of size-1 dimensions as an explicit broadcast.

**Comment by jjsjann123** [comment]
> 1. Could we avoid fusing `expand` if it is a fusion output, since we will need to return an aliased tensor with stride 0?

We can be a little more conservative. I think we can check that the output of expand can have only one consumer that is the `CudaFusionGroup`.

> 3. Treat `expand` as a no-op in the fusion because we're using expand of size-1 dimensions as an explicit broadcast.

no-op meaning we don't mark the broadcasted dimensions? That would give us wrong output sizes if we ended up using the tensor without broadcast (or having something else broadcasted to the target tensor).
e.g.
```
a = x.expand(...)
return a.relu()
```

I think we can output a tensor with broadcasted dimension with an implicit size-1???

**Comment by rdspring1** [comment]
For unary operations, we could partially-resolve the expanded dimensions.

```python
x = torch.randn(10, 1)
a = x.expand(-1, 10)
output = a.relu()
```

NvFuser TensorView:
```python
x [10, 1-broadcast]
# The expanded dimension is marked as broadcast with concrete size 10. 
a [10, 10-broadcast, expand]
output [10, 10]
```

**Comment by jjsjann123** [comment]
Sorry I never posted my comment yesterday and also that I should be more concrete about my example :bow: 

I wasn't considering `expand` to a non-trivially-broadcasted size (!=1) yet. I was merely referring to the previous suggestion of treating `expand` as no-op for size-1.

e.g.
```
x = torch.randn(10, 2)
a = x.expand(1, 10, 2)
output = a.relu()
```

Same as your example above, we can't treat `expand` with size-1 as no-op, but have to explicitly mark the first dimension as broadcasted. Otherwise, `output` would be of shape `(10, 2)` instead of `(1, 10, 2)`

**Comment by jjsjann123** [comment]
Ah, I think I was reading the request wrong.... I thought we are only expecting the expanded dimension to be size-1... But that doesn't look like what the request was asking for.

Ryan's example makes a lot more sense to me now.

If we just look at this section in the example:
```
        out1 = inputs * 2.                                                                                                                                                                                                                           
        out2 = out1.sum(dim=-1, keepdim=True)                                                                                                                                                                                                        
        out3 = out2.expand((2, 4, 6))                                                                                                                                                                                                                                                                                                                                                                                                                                             
        out4 = out3 * 3.                                                                                                                                                                                                                             
        out5 = out4 * 5.                                                                                                                                                                                                                             
        return out5
```
I think Kevin was asking us to support: Expanding on fixed dimensions with static sizes, while the rest of the dimensions that are not expanded could be dynamic.
So this sounds similar to our `view` guard logic does, thoughts? (soft tag @rdspring1 ).

FYI, regarding the aliasing issue, I think we can cheat that by using the same alias analysis regarding view ops. (updating expand to expand_copy and target that in our fusion logic).

**Comment by rdspring1** [comment]
I think we can handle your case where we would add new broadcast dimensions to the tensor.

Option 1:
Enforce a static size constraint to expanded dimensions. The -1 dimensions are dynamic. <--- Similar to view guard logic

Option 2:
We check that the expanded dimension size matches the resolved broadcast dimension size in downstream operations.

e.g., If the expanded dimension is resolved in the fusion, we would enforce that the `expand` dim-1 size matches the dim-1 size for y.
```python
x = torch.randn(10, 1) # [10, 1-broadcast]
y = torch.rand(10, 20) # [10, 20]
a = x.expand(-1, 20) # [10, 20-broadcast]
return output = x + y #[10, 20]
```

**Comment by jjsjann123** [comment]
> Option 1: Enforce a static size constraint to expanded dimensions. The -1 dimensions are dynamic. <--- Similar to view guard logic
> Option 2: We check that the expanded dimension size matches the resolved broadcast dimension size in downstream operations.

Both sounds fine to me regarding expanded size-1 dimension to be parsed as broadcast. I'm slightly leaning towards option 1 because it uses existing logic :)

Regarding option 2: What if we don't have a `resolved broadcast dimension size`. e.g. expanded dimension doesn't get `merged` to a normal dimension from a different tensor.
```
        out1 = inputs * 2.                                                                                                                                                                                                                           
        out2 = out1.sum(dim=-1, keepdim=True)                                                                                                                                                                                                        
        out3 = out2.expand((2, 4, 6))                                                                                                                                                                                                                                                                                                                                                                                                                                             
        out4 = out3 * 3.                                                                                                                                                                                                                             
        out5 = out4 * 5.                                                                                                                                                                                                                             
        return out5
```
You mentioned something like this earlier vvv
```
# The expanded dimension is marked as broadcast with concrete size 10. 
a [10, 10-broadcast, expand]
```
Is there an way to do that natively on a TensorView?
Otherwise, we'll have to hack it by faking a no-op elem-wise along a fake tensor, and we'll only be able to support static `expanded` dimension, but this should still satisfy Kevin's ask, since it only requires dynamic size on `non-expanded` dimensions.

**Comment by jjsjann123** [comment]
We now surprisingly have proper expand support in nvfuser (for static expand sizes). We should close this one.

Feel free to reopen if we needed dynamic shape support. That should be do-able, we just need to plumb vector of int in our executor.

**Comment by jjsjann123** [comment]
linking #1722 #1739 #1754
